cluster/src/main/java/org/apache/iotdb/cluster/log/UnCommittedEntryManager.java

```diff
@@ -0,0 +1,149 @@
+package org.apache.iotdb.cluster.log;
+
+import org.apache.iotdb.cluster.exception.EntryStabledException;
+import org.apache.iotdb.cluster.exception.EntryUnavailableException;
+import org.apache.iotdb.cluster.exception.GetEntriesWrongParametersException;
+import org.apache.iotdb.cluster.exception.TruncateCommittedEntryException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class UnCommittedEntryManager {
+    private static final Logger logger = LoggerFactory.getLogger(UnCommittedEntryManager.class);
+    // all entries that have not been committed.
+    private List<Log> entries;
+    // the first uncommitted entry index.
+    private long offset;
+
+    public UnCommittedEntryManager(long offset) {
+        this.offset = offset;
+        this.entries = new ArrayList<>();
+    }
+
+    /**
+     * getFirstUnCommittedIndex return the first uncommitted index.
+     *
+     * @return offset.
+     */
+    public long getFirstUnCommittedIndex() {
+        return offset;
+    }
+
+
+    /**
+     * maybeLastIndex returns the last index if it has at least one uncommitted entry.
+     *
+     * @return set to -1 if unstable entries are empty, or return the last index.
+     */
+    public long maybeLastIndex() {
+        int entryNum = entries.size();
+        if (entryNum != 0) {
+            return offset + entryNum - 1;
+        }
+        return -1;
+    }
+
+    /**
+     * maybeTerm returns the term for given index.
+     *
+     * @param index
+     * @return set to -1 if the entry for this index cannot be found, or return the entry's term.
+     */
+    public long maybeTerm(long index) {
+        if (index < offset) {
+            return -1;
+        }
+        long last = maybeLastIndex();
+        if (last == -1 || index > last) {
+            return -1;
+        }
+        return entries.get((int) (index - offset)).getCurrLogTerm();
+    }
+
+    /**
+     * stableTo remove useless prefix entries as long as these entries has been committed,
+     * It is often called after snapshot persistence and log persistence.
+     *
+     * @param index
+     * @param term
+     */
+    public void stableTo(long index, long term) {
+        long entryTerm = maybeTerm(index);
+        // only update the uncommitted entries if term is matched with
+        // an uncommitted entry.
+        if (entryTerm == term) {
+            entries.subList(0, (int) (index + 1 - offset)).clear();
+            offset = index + 1;
+        }
+    }
+
+    /**
+     * truncateAndAppend append uncommitted entries.It will truncate conflict entries if it find inconsistencies.
+     *
+     * @param entries
+     * @throws TruncateCommittedEntryException
+     */
+    public void truncateAndAppend(List<Log> entries) throws TruncateCommittedEntryException {
```

Comment:
Mention that the caller should deal with cases when `len > this.entries.size()`.

Reply:
I will throw a exception for this.

Reply:
Yes, but it would be better to clear it out in the JavaDoc, in case someone else misuse it.

Reply:
OK,I have added JavaDoc.